packages/muelu/src/Transfers/Smoothed-Aggregation/MueLu_SaPFactory_kokkos_def.hpp

```diff
@@ -360,6 +360,207 @@ struct constraintKernel {
 
    }
 };
+
+template<typename local_matrix_type>
+struct optimalSatisfyConstraintsForScalarPDEsKernel {
+
+   using Scalar= typename local_matrix_type::non_const_value_type;
+   using SC=Scalar;
+   using LO=typename local_matrix_type::non_const_ordinal_type;
+   using Device= typename local_matrix_type::device_type;
+   const Scalar zero = Kokkos::ArithTraits<SC>::zero();
+   const Scalar one  = Kokkos::ArithTraits<SC>::one();
+   LO     nPDEs;
+   local_matrix_type localP;
+   Kokkos::View<SC**, Device> origSorted;
+   Kokkos::View<SC**, Device> fixedSorted;
+   Kokkos::View<LO**, Device> inds;
+
+   optimalSatisfyConstraintsForScalarPDEsKernel(LO nPDEs_, LO maxRowEntries_, local_matrix_type localP_) : nPDEs(nPDEs_), localP(localP_)
+   {
+     origSorted  =  Kokkos::View<SC**, Device>("origSorted" , localP_.numRows(), maxRowEntries_);
+     fixedSorted =  Kokkos::View<SC**, Device>("fixedSorted", localP_.numRows(), maxRowEntries_);
+     inds        =  Kokkos::View<LO**, Device>("inds"       , localP_.numRows(), maxRowEntries_);
+   }
+
+   KOKKOS_INLINE_FUNCTION
+   void operator() (const size_t rowIdx) const {
+
+     auto rowPtr = localP.graph.row_map;
+     auto values = localP.values;
+
+     auto nnz = rowPtr(rowIdx+1) - rowPtr(rowIdx);
+
+     if (nnz != 0) {
+
+       Scalar rsumTarget = zero;
+       for (auto entryIdx = rowPtr(rowIdx); entryIdx < rowPtr(rowIdx + 1); entryIdx++) rsumTarget += values(entryIdx);
+
+       {
+         SC aBigNumber;
+         SC rowSumDeviation, temp, delta;
+         SC closestToLeftBoundDist, closestToRghtBoundDist;
+         LO closestToLeftBound, closestToRghtBound;
+         bool   flipped;
+
+         SC leftBound = zero;
+         SC rghtBound = one;
+         if ((Kokkos::ArithTraits<SC>::real(rsumTarget) >= Kokkos::ArithTraits<SC>::real(leftBound*((SC)nnz))) && 
+             (Kokkos::ArithTraits<SC>::real(rsumTarget) <= Kokkos::ArithTraits<SC>::real(rghtBound*((SC)nnz)))){ // has Feasible solution
+
+           flipped    = false;
+           // compute aBigNumber to handle some corner cases where we need
+           // something large so that an if statement will be false
+           aBigNumber = Kokkos::ArithTraits<SC>::zero();
+           for (auto entryIdx = rowPtr(rowIdx); entryIdx < rowPtr(rowIdx + 1); entryIdx++){
+             if ( Kokkos::ArithTraits<SC>::magnitude( values(entryIdx) ) > Kokkos::ArithTraits<SC>::magnitude(aBigNumber)) 
+               aBigNumber = Kokkos::ArithTraits<SC>::magnitude( values(entryIdx) );
+           }
+           aBigNumber = aBigNumber+ (Kokkos::ArithTraits<SC>::magnitude(leftBound) + Kokkos::ArithTraits<SC>::magnitude(rghtBound))*((SC)100.0);
```

Comment:
Instead of ((SC)100.0) you should do `100*KAT::one()` that way you will likely not get any conversion failures